src/core/mount.c

```diff
@@ -352,7 +352,10 @@ static int mount_add_device_dependencies(Mount *m) {
          * automatically stopped when the device disappears suddenly. */
         dep = mount_is_bound_to_device(m) ? UNIT_BINDS_TO : UNIT_REQUIRES;
 
-        mask = m->from_fragment ? UNIT_DEPENDENCY_FILE : UNIT_DEPENDENCY_MOUNTINFO_IMPLICIT;
+        /* We update 'what' based on /proc/self/mountinfo, so let's say
+         * the dependency is based on that, even if we have a fragment. */
+        mask = m->from_fragment && !m->from_proc_self_mountinfo ? UNIT_DEPENDENCY_FILE
+                                                                : UNIT_DEPENDENCY_MOUNTINFO_IMPLICIT;
```

Comment:
hmm, actually i think the check should be this:

```c
mask = m->from_proc_self_mountinfo ? UNIT_DEPENDENCY_MOUNTINFO_IMPLICIT : UNIT_DEPENDENCY_FILE;
```

Why? because this is then in sync with the origin of `p`, see definition of get_mount_parameters()....

or in other words, I don't think we should test m->from_fragment here at all..

Reply:
Yes, that's much better.